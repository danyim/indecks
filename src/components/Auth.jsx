import React from 'react'
import PropTypes from 'prop-types'
import SignUpLoginForm from '../components/SignUpLoginForm'
import styles from '../styles/components/Auth.styl'

class Auth extends React.Component {
  static propTypes = {
    authenticated: PropTypes.bool,
    saveDecksToFirebase: PropTypes.func.isRequired,
    displayName: PropTypes.string,
    email: PropTypes.string,
    isAuthenticating: PropTypes.bool,
    logout: PropTypes.func.isRequired,
    fetchUserDecks: PropTypes.func.isRequired,
    login: PropTypes.func.isRequired,
    signUpEmail: PropTypes.func.isRequired,
    signInGithub: PropTypes.func.isRequired, // eslint-disable-line
    signInTwitter: PropTypes.func.isRequired, // eslint-disable-line
    signInGoogle: PropTypes.func.isRequired, // eslint-disable-line
    // photoURL: PropTypes.string,
  }

  static defaultProps = {
    isAuthenticating: false,
    authenticated: false,
    displayName: null,
    email: null,
    // photoURL: null,
  }

  render() {
    return (
      <div className={styles.auth}>
        {!this.props.authenticated ? (
          <SignUpLoginForm
            isAuthenticating={this.props.isAuthenticating}
            login={this.props.login}
            signUpEmail={this.props.signUpEmail}
            signInGithub={this.props.signInGithub}
            signInTwitter={this.props.signInTwitter}
            signInGoogle={this.props.signInGoogle}
          />
        ) : (
          <div>
            <p>
              {/*
                this.props.photoURL
                ? <img src={this.props.photoURL} alt='Avatar' />
                : null
              */}
              Logged in as&nbsp;
              <strong>{this.props.displayName ? this.props.displayName : this.props.email}</strong>
            </p>
            <button type="button" className="btn" onClick={this.props.saveDecksToFirebase}>
              Save your decks
            </button>
            <button type="button" className="btn" onClick={this.props.fetchUserDecks}>
              Reload decks from account
            </button>
            <button type="button" className="button outline" onClick={this.props.logout}>
              Log out
            </button>
          </div>
        )}
      </div>
    )
  }
}

export default Auth
