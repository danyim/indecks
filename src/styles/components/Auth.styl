@import '../_variables'

// figure is added here to give precedence over the base grid-figure
figure.grid-figure
  display: flex
  min-height: 275px
  flex-flow: column nowrap
  justify-content: center

.auth
  margin: 2rem 0
  strong
    display: inline
  > div
    flex-flow: column nowrap

.control-buttons
  flex: 0 0 auto
  flex-flow: row nowrap
  margin-top: 1.5rem
  button
    margin: 0 0 0 1rem
  button:first-child
    margin: 0

@media (max-width: cardWidth)
  .control-buttons
    flex: 0 0 50%
    flex-flow: column nowrap
    align-content: space-around
    button
      margin: .6rem 0 0
    button:first-child
      margin: 0
