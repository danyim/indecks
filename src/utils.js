export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

