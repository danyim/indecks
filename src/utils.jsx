import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
