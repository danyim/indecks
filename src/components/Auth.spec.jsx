import toJSON from 'enzyme-to-json'
import 'jest-styled-components' // eslint-disable-line
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import Auth from './Auth'

describe('Auth', () => {
  let props

  beforeEach(() => {
    props = {
      authenticated: false,
      displayName: '',
      email: '',
      isAuthenticating: false,
      logout: () => {},
      saveDecksToFirebase: () => {},
      fetchUserDecks: () => {},
      login: () => {},
      signUpEmail: () => {},
      signInGithub: () => {},
      signInTwitter: () => {},
      signInGoogle: () => {},
    }
  })

  it('should render self and subcomponents', () => {
    const { wrapper } = mount(Auth, props)
    const tree = toJSON(wrapper)
    expect(tree).toMatchSnapshot()
  })

  it('should show the signup/login form when not authenticated', () => {
    const { wrapper } = setupFull(Auth, {
      ...props,
      authenticated: false,
    })

    const form = wrapper.find('SignUpLoginForm')

    expect(form.exists()).toEqual(true)
  })

  it('should display the username or email when authenticated', () => {
    const { wrapper } = setupFull(Auth, {
      ...props,
      authenticated: true,
    })

    const subject = wrapper.find('p')

    expect(subject.exists()).toEqual(true)
    expect(subject.text()).toContain('Logged in as')
  })

  it('should display the displayName if available', () => {
    const { wrapper } = setupFull(Auth, {
      ...props,
      authenticated: true,
      displayName: 'test',
    })

    const subject = wrapper.find('strong')

    expect(subject.exists()).toEqual(true)
    expect(subject.text()).toContain('test')
  })

  it("should display the email if displayName isn't available", () => {
    const { wrapper } = setupFull(Auth, {
      ...props,
      authenticated: true,
      displayName: null,
      email: 'test@gmail.com',
    })

    const subject = wrapper.find('strong')

    expect(subject.exists()).toEqual(true)
    expect(subject.text()).toContain('test@gmail.com')
  })

  // it('should call handleClick when any selection is clicked', () => {
  //   const handler = jest.fn()
  //   const { wrapper } = setupFull(Auth, {
  //     ...props,
  //     handleClick: handler
  //   })

  //   const buttonLeft = wrapper.find('button').last()
  //   const buttonRight = wrapper.find('button').last()

  //   buttonLeft.simulate('click')
  //   expect(handler).toHaveBeenCalled()
  //   buttonRight.simulate('click')
  //   expect(handler).toHaveBeenCalledTimes(2)
  // })
})
