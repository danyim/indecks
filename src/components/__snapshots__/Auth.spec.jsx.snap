// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth should render self and subcomponents 1`] = `undefined`;
